blog: wasmCloud 2.5 release blog (#1210)
wasmCloud/wasmCloud.com: blog/2026-06-30-wasmcloud-2.5.0/index.mdx

One implementation note documented in the same release: WASI 0.3 always brings the `component-model-async` proposal along with it, so every 2.5.0 host has the async proposal available whether or not it's listed explicitly in the proposals set. 

## Multiplexing host plugins with the WebAssembly Component Model's `implements` clause

The most user-visible payoff of this release is finally being able to do this in a WIT world:

```wit
package my-app:kv-cached;

world kv-cached {
    import cache: wasi:keyvalue/store@0.2.0-draft;
    import store: wasi:keyvalue/store@0.2.0-draft;
    export wasi:http/incoming-handler@0.2.2;
}
```

That's two labeled imports of the same `wasi:keyvalue/store` interface, expressed via the Component Model's `implements` clause (the `import <label>: <iface>;` syntax). The guest now has control over which store each call hits: `cache::open("session-data")` and `store::open("orders")` reach two different backends (say, a fast local cache and a durable NATS-backed store) without the guest needing to know which is which.

:::info[Feature flag required for now]

> `(implements ..)` multiplexing ships behind the `wasm_component_model_implements` Cargo feature, which is not in the default `wash-runtime` feature set. That means the stock `ghcr.io/wasmcloud/wash:2.5.0` binary and the stock `ghcr.io/wasmcloud/runtime-operator:2.5.0` image don't enable multi-backend binding out of the box. To use it today, build a custom host image with `CARGO_FEATURES=wasm_component_model_implements` (the source Dockerfile takes this as a build arg) and point the chart at it via [`runtime.image.tag`](/docs/kubernetes-operator/operator-manual/helm-values/#runtimeimagetag). The `implements` proposal is a WebAssembly Component Model feature being finalized upstream; once it graduates and lands in a tagged Wasmtime release, the workspace fork can go and this can move onto the default surface.

Reviewer: ```suggestion
`(implements ..)` multiplexing ships behind the `wasm_component_model_implements` Cargo feature, which is not in the default `wash-runtime` feature set. That means the stock `ghcr.io/wasmcloud/wash:2.5.0` binary and the stock `ghcr.io/wasmcloud/runtime-operator:2.5.0` image don't enable multi-backend binding out of the box. To use it today, build a custom host image with `CARGO_FEATURES=wasm_component_model_implements` (the source Dockerfile takes this as a build arg) and point the chart at it via [`runtime.image.tag`](/docs/kubernetes-operator/operator-manual/helm-values/#runtimeimagetag). The `implements` proposal is a WebAssembly Component Model feature being finalized upstream; once it graduates and lands in a tagged Wasmtime release, wasmCloud will also enable the feature by default.
```

* **Top-level Wasm proposals are now runtime-configurable** per host, per host group, or per `wash dev` config
* **`(implements ..)` have landed** (behind a feature flag for now), meaning a workload can now import the same interface multiple times under distinct labels and route each call to its own backend.

2.5.0 also ships async-shaped `wasmcloud:keyvalue` and `wasmcloud:blobstore` WIT (with TTL, CAS, and richer errors), a Helm chart hardening pass with NetworkPolicies on the runtime-operator pods, an operator host-controller refactor to Server-Side Apply, and a passel of `wash` quality-of-life fixes.

{/* truncate */}

## At a glance

| Area | 2.4.0 | 2.5.0 |
|---|---|---|
| Wasmtime | 45 | 46 |
| WASI default | 0.2 | 0.3 |
| Engine Wasm proposals | compile-time | runtime-configurable per host |
| Host-plugin routing | one backend per interface | multiplexed via `(implements ..)` (feature-flagged) |
| Async `wasmcloud:keyvalue` / `wasmcloud:blobstore` WIT | not shipped | shipped (TTL, CAS, typed errors) |
| Helm chart NetworkPolicies | not shipped | runtime-operator and host group pods |

## Wasmtime 46 and WASI 0.3 by default

> Wasmtime 46 brings the engine refactor that makes this release's other headlines possible. WASI 0.3 is now on by default across the wash binary: `--wasip3` / `WASH_WASIP3` and `DevConfig.wasip3` all default to `enabled`. If you'd like to opt out for whatever reason, you can do so with `--wasip3 false` or `wasip3: false` in `.wash/config.yaml`.

Reviewer: The flag should actually be removed and there's no opt-out

## Top-level Wasm proposals as a configurable surface

The engine refactor that came with the Wasmtime 46 bump made builder settings (pooling, fuel, proposals) composable on top of any `with_config()` base instead of erroring when combined. That refactor is what made it possible to expose the top-level Wasm proposals (`component-model-async`, `gc`, `exception-handling`, `wide-arithmetic`, `threads`, `tail-call`) as a per-host, per-host-group surface rather than a baked-in choice.

The new API is a `WasmProposal` enum with `with_wasm_proposal()` on the builder ([#5267](https://github.com/wasmCloud/wasmCloud/pull/5267), driven by maintainer [@Aditya1404Sal](https://github.com/Aditya1404Sal)). From a user standpoint:

- **`wash host`** accepts `--wasm-proposal <name>` (comma-separated, repeatable, also `WASH_WASM_PROPOSALS`).
- **`wash dev`** reads a `dev.wasm_proposals` list out of `.wash/config.yaml`, validated at startup.
- **`runtime-operator` Helm chart** accepts a per-host-group `wasmProposals: []` list that the chart renders as `--wasm-proposal` arguments on the host container. The chart's appVersion bumps to 0.6.0 in step with the new field.

> One implementation note documented in the same release: `with_wasip3()` implies `component-model-async`, so a host group that opts into WASI 0.3 has the async proposal whether or not it's listed explicitly. 

Reviewer: no opt-in

## Multiplexing host plugins with the WebAssembly Component Model's `implements` clause

The most user-visible payoff of this release is finally being able to do this in a WIT world:

```wit
package my-app:multi-tenant;

world multi-tenant {
    import team-a: wasi:keyvalue/store@0.2.0-draft;
    import team-b: wasi:keyvalue/store@0.2.0-draft;
    export wasi:http/incoming-handler@0.2.2;
}
```

> That's two labeled imports of the same `wasi:keyvalue/store` interface, expressed via the Component Model's `implements` clause (the `import <label>: <iface>;` syntax). Inside the guest, each label surfaces as its own import namespace — `team_a::open("bucket")` and `team_b::open("bucket")` reach two different backends (say, Redis and NATS) without the guest having to know which is which.

Reviewer: I'd probably say the guest now has control over being able to use multiple stores (I prefer a local cache vs store vs the teams example)

:::info[Feature flag required for now]

> `(implements ..)` multiplexing ships behind the `wasm_component_model_implements` Cargo feature, which is not in the default `wash-runtime` feature set. That means the stock `ghcr.io/wasmcloud/wash:2.5.0` binary and the stock `ghcr.io/wasmcloud/runtime-operator:2.5.0` image don't enable multi-backend binding out of the box. To use it today, build a custom host image with `CARGO_FEATURES=wasm_component_model_implements` (the source Dockerfile takes this as a build arg) and point the chart at it via [`runtime.image.tag`](/docs/kubernetes-operator/operator-manual/helm-values/#runtimeimagetag). The `implements` proposal is being tracked upstream in Wasmtime; once it lands in a tagged release, the feature will move onto the default surface.

Reviewer: tracked upstream as a component model feature

The async-keyvalue world also picks up an `import types` line as part of the `bucket` restructure, so the resource type stays identical across `store`, `cas`, `atomics`, `batch`, and `watcher`.

## Helm chart hardening

The `runtime-operator` Helm chart picks up two security and operability improvements in 2.5.0:

- **NetworkPolicies for runtime-operator pods** ([#5269](https://github.com/wasmCloud/wasmCloud/pull/5269)) — the operator and host group pods now ship with default-allow-only-what-they-need NetworkPolicies, closing the implicit "the cluster network is open" assumption that earlier releases carried.
- **Chart lint and harden pass** ([#5270](https://github.com/wasmCloud/wasmCloud/pull/5270)) — broader hardening of resource limits, security contexts, probe defaults, and chart-lint compliance across the chart's templates.

If you've been hand-patching NetworkPolicies onto a wasmCloud release, this release replaces that with the in-chart equivalent.

> ## WIT publishing and CI

Reviewer: remove this section

Two CI changes worth knowing about for anyone consuming wasmCloud's WIT packages directly:

- [WIT is now published to a namespaced OCI path](https://github.com/wasmCloud/wasmCloud/pull/5253) (`ghcr.io/wasmcloud/<namespace>/<package>`) so `wkg` and `wash wit` can resolve packages by their canonical reference.
- [WIT package builds use `wkg` instead of `wash`](https://github.com/wasmCloud/wasmCloud/pull/5252), aligning the publish path with how downstream consumers resolve dependencies.

## Other notable changes

- **[Server-Side Apply / Upsert in the host controller](https://github.com/wasmCloud/wasmCloud/pull/5226).** Replaces the cache-get-modify-write pattern with Server-Side Apply, eliminating a class of conflict-and-retry behavior the operator used to fall into on contended hosts. Paired with `Always set leader-elect` ([#5225](https://github.com/wasmCloud/wasmCloud/pull/5225)) and a fix to only update `LastSeen` on metadata change ([#5298](https://github.com/wasmCloud/wasmCloud/pull/5298)) for less reconcile noise.
- **[s390x binary in CI](https://github.com/wasmCloud/wasmCloud/pull/5249).** wasmCloud now ships a Linux/s390x binary alongside the existing platform set.

> - **[wasi:webgpu back on by default](https://github.com/wasmCloud/wasmCloud/pull/5272).** The Wasmtime 46 bump initially moved wasi-webgpu behind an opt-in feature; #5272 puts it back into the default set now that the compatibility constraint with wasmtime 45 is resolved.

Reviewer: this happened internal to the release (was never exposed in a release) so drop
